contrib/src/main/java/com/datatorrent/contrib/cassandra/AbstractUpsertOutputOperator.java

```diff
@@ -0,0 +1,1401 @@
+ * 6. Support for custom mapping of POJO payload field names to that of cassandra columns. Practically speaking,
+ *    POJO field names might not always match with Cassandra Column names and hence this support. This will also avoid
+ *    writing a POJO just for the cassandra operator and thus an existing POJO can be passed around to this operator.
+ *    Please refer javadoc {@link this.getPojoFieldNameToCassandraColumnNameOverride()} for an example
+ * 7. TTL support - A default TTL can be set for the Connection ( via {@link ConnectionStateManager} and then used
+ *    for all mutations. This TTL can further be overridden at a tuple execution level to accomodate use cases of
+ *    setting custom column expirations typically useful in wide row implementations.
+ * 8. Support for Counter Column tables. Counter tables are also supported with the values inside the incoming
+ *    POJO added/subtracted from the counter column accordingly. Please note that the value is not absolute set but
+ *    rather representing the value that needs to be added to or subtracted from the current counter.
+ * 9. Support for Composite Primary Keys is also supported. All the POJO fields that map to the composite
+ *    primary key are used to resolve the primary key in case of a Composite Primary key table
+ * 10. Support for conditional updates : This operator can be used as an Update Only operator as opposed to an
+ *     Upsert operator. i.e. Update only IF EXISTS . This is achieved by setting the appropriate boolean in the
+ *     {@link UpsertExecutionContext} tuple that is passed from the upstream operator.
+ * 11. Lenient mapping of POJO fields to Cassandra column names. By default the POJO field names are case insensitive
+ *     to cassandra column names. This can be further enhanced by over-riding mappings. Please refer feature 6 above.
+ * 12. Defaults can be overridden at at tuple execution level for TTL & Consistency Policies
+ * 13. Support for handling Nulls i.e. whether null values in the POJO are to be persisted as is or to be ignored so
+ *     that the application need not perform a read to populate a POJO field if it is not available in the context
+ * 14. A few autometrics are provided for monitoring the latency aspects of the cassandra cluster
+ */
+
+@InterfaceStability.Evolving
+public abstract class AbstractUpsertOutputOperator extends BaseOperator implements
+    Operator.ActivationListener<Context.OperatorContext>, Operator.CheckpointNotificationListener
+{
+
+  public static final String TTL_PARAM_NAME = "ttl";
+
+  ConnectionStateManager connectionStateManager;
+
+  transient WindowDataManager windowDataManager;
+
+  private long currentWindowId;
+
+  private transient boolean isInSafeMode;
+
+  private transient long reconcilingWindowId;
+
+  private transient boolean isInReconcilingMode;
+
+  transient Session session;
+
+  transient Cluster cluster;
+
+  transient String keyspaceName;
```

Comment:
will make the keyspacename and table name as private. Marked them transient only to avoid unnecessary state passivation .. My understanding was that activate() is called for all situations i.e. when an operator comes back alive after restarts/crashes. I was hoping to make the ConnectionStatemanager passivated state so that the all aspects like reconnection times etc which are not exposed as variables inside the operator itself can be extracted from the connection state manager and a new session established ...  ( and this also makes the session object inside the clustermanager a transient aspect of the passivation state )

Please let me know if my understanding of activate is wrong here and will modify the code aspects accordingly.